Mac failure is:
```
/Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swiftpm/Sources/Build/SPMSwiftDriverExecutor.swift:80:39: error: value of type '[Job.ArgTemplate]' has no member 'joinedArguments'
10:51:18       result += " # \(job.commandLine.joinedArguments)"
10:51:18                       ~~~~~~~~~~~~~~~ ^~~~~~~~~~~~~~~
10:51:18 [14/21][ 66%][34.489s] Linking Swift shared library lib/libXcodeproj.dylib
10:51:18 ninja: build stopped: subcommand failed.
```

--- a/lib/Sema/TypeCheckConstraints.cpp
+++ b/lib/Sema/TypeCheckConstraints.cpp
@@ -592,17 +592,9 @@ bool TypeChecker::typeCheckForEachBinding(DeclContext *dc, ForEachStmt *stmt) {
   // check to see if the sequence expr is throwing (and async), if so require 
   // the stmt to have a try loc
   if (stmt->getAwaitLoc().isValid()) {
-    auto Ty = sequence->getType();
-    if (Ty.isNull()) { 
-      auto DRE = dyn_cast<DeclRefExpr>(sequence);
-      if (DRE) {
-        Ty = DRE->getDecl()->getInterfaceType();
-      }
-      if (Ty.isNull()) {
-        return failed();
-      }
-    }
-    
+    // fetch the sequence out of the statement
+    // else wise the value is potentially unresolved
+    auto Ty = stmt->getSequence()->getType();
     auto module = dc->getParentModule();
     auto conformanceRef = module->lookupConformance(Ty, sequenceProto);
     

--- a/lib/Sema/TypeCheckEffects.cpp
+++ b/lib/Sema/TypeCheckEffects.cpp
@@ -528,6 +528,13 @@ public:
       }
 
       if (classifiedAsThrows) {
+        // multiple passes can occur, so ensure that any sub-expressions of this
+        // call are marked as throws to mimic the closure variant.
+        if (auto subExpr = dyn_cast<ApplyExpr>(E->getFn())) {
+          if (!subExpr->isThrowsSet()) {
+            subExpr->setThrows(true);
+          }
+        }
         return Classification::forRethrowingOnly(
           PotentialThrowReason::forRethrowsConformance(E), isAsync);
       }

--- a/test/Concurrency/async_sequence_syntax.swift
+++ b/test/Concurrency/async_sequence_syntax.swift
@@ -32,3 +32,26 @@ func missingAsyncInBlock<T : AsyncSequence>(_ seq: T) {
     } catch { }
   }
 }
+
+func doubleDiagCheckGeneric<T : AsyncSequence>(_ seq: T) async {
+  var it = seq.makeAsyncIterator()
+  // expected-note@+2{{call is to 'rethrows' function, but a conformance has a throwing witness}}
+  // expected-error@+1{{call can throw, but it is not marked with 'try' and the error is not handled}}
+  let _ = await it.next()
+}
+
+struct ThrowingAsyncSequence: AsyncSequence, AsyncIteratorProtocol {
+  typealias Element = Int
+  typealias AsyncIterator = Self
+  mutating func next() async throws -> Int? {
+    return nil
+  }
+
+  func makeAsyncIterator() -> Self { return self }
+}
+
+func doubleDiagCheckConcrete(_ seq: ThrowingAsyncSequence) async {
+  var it = seq.makeAsyncIterator()
+  // expected-error@+1{{call can throw, but it is not marked with 'try' and the error is not handled}}
+  let _ = await it.next()
+}

